src/gfn/gym/graph_building.py

```diff
@@ -0,0 +1,258 @@
+        else:
+            add_edge_states = states[add_edge_mask].tensor
+            add_edge_actions = actions[add_edge_mask]
+            if torch.any(
+                add_edge_actions.edge_index[:, 0] == add_edge_actions.edge_index[:, 1]
+            ):
+                return False
+            if add_edge_states["node_feature"].shape[0] == 0:
+                return False
+            if torch.any(
+                add_edge_actions.edge_index > add_edge_states["node_feature"].shape[0]
+            ):
+                return False
+            global_edge_index = (
+                add_edge_actions.edge_index + add_edge_states["batch_ptr"][:-1][:, None]
+            )
+            equal_edges_per_batch = torch.all(
+                add_edge_states["edge_index"] == global_edge_index[:, None], dim=-1
+            ).sum(dim=-1)
+
+            if backward:
+                add_edge_out = torch.all(equal_edges_per_batch == 1)
+            else:
+                add_edge_out = torch.all(equal_edges_per_batch == 0)
+
+        return bool(add_node_out) and bool(add_edge_out)
+
+    def _add_node(
+        self,
+        tensor_dict: TensorDict,
+        batch_indices: torch.Tensor,
+        nodes_to_add: torch.Tensor,
+    ) -> TensorDict:
+        if isinstance(batch_indices, list):
+            batch_indices = torch.tensor(batch_indices)
+        if len(batch_indices) != len(nodes_to_add):
+            raise ValueError(
+                "Number of batch indices must match number of node feature lists"
+            )
+
+        modified_dict = tensor_dict.clone()
+        node_feature_dim = modified_dict["node_feature"].shape[1]
+
+        for graph_idx, new_nodes in zip(batch_indices, nodes_to_add):
+            tensor_dict["batch_ptr"][graph_idx]
+            end_ptr = tensor_dict["batch_ptr"][graph_idx + 1]
+
+            if new_nodes.ndim == 1:
+                new_nodes = new_nodes.unsqueeze(0)
+            if new_nodes.shape[1] != node_feature_dim:
+                raise ValueError(
+                    f"Node features must have dimension {node_feature_dim}"
+                )
+
+            # Update batch pointers for subsequent graphs
+            shift = new_nodes.shape[0]
+            modified_dict["batch_ptr"][graph_idx + 1 :] += shift
+
+            # Expand node features
+            modified_dict["node_feature"] = torch.cat(
+                [
+                    modified_dict["node_feature"][:end_ptr],
+                    new_nodes,
+                    modified_dict["node_feature"][end_ptr:],
+                ]
+            )
+
+            # Update edge indices
+            # Increment indices for edges after the current graph
+            edge_mask_0 = modified_dict["edge_index"][:, 0] >= end_ptr
+            edge_mask_1 = modified_dict["edge_index"][:, 1] >= end_ptr
+            modified_dict["edge_index"][edge_mask_0, 0] += shift
+            modified_dict["edge_index"][edge_mask_1, 1] += shift
```

Comment:
I think this looks right. This is note to self to ensure we have solid tests on this logic (and the backward step of removing).

Reply:
Yes, the test_graph_env in test_environments is supposed to smoke test this as well